api/task_manager.cc

```diff
@@ -0,0 +1,130 @@
+
+#include <utility>
+
+namespace api {
+
+namespace tm = httpd::task_manager_json;
+using namespace json;
+
+inline bool filter_tasks(task_manager::task_ptr task, std::unordered_map<sstring, sstring>& query_params) {
+    return (!query_params.contains("keyspace") || query_params["keyspace"] == task->get_status().keyspace) &&
+        (!query_params.contains("table") || query_params["table"] == task->get_status().table);
+}
+
+struct full_task_status {
+    task_manager::task::status task_status;
+    std::string module;
+    task_id parent_id;
+    is_abortable abortable;
+};
+
+tm::task_status make_status(full_task_status status) {
+    auto start_time = db_clock::to_time_t(status.task_status.start_time);
+    auto end_time = db_clock::to_time_t(status.task_status.end_time);
+    ::tm st, et;
+    ::gmtime_r(&end_time, &et);
+    ::gmtime_r(&start_time, &st);
+
+    tm::task_status res{};
+    res.id = status.task_status.id.to_sstring();
+    res.type = status.task_status.type;
+    res.state = status.task_status.state;
+    res.is_abortable = bool(status.abortable);
+    res.start_time = st;
+    res.end_time = et;
+    res.error = status.task_status.error;
+    res.parent_id = status.parent_id.to_sstring();
+    res.number = status.task_status.number;
+    res.shard = status.task_status.shard;
+    res.keyspace = status.task_status.keyspace;
+    res.table = status.task_status.table;
+    res.entity = status.task_status.entity;
+    res.progress_units = status.task_status.progress_units;
+    res.progress_total = status.task_status.progress.total;
+    res.progress_completed = status.task_status.progress.completed;
+    return res;
+}
+
+future<json::json_return_type> retrieve_status(task_manager::foreign_task_ptr task) {
+    if (task.get() == nullptr) {
+        return make_exception_future<json::json_return_type>(httpd::bad_param_exception("Invalid task id"));
+    }
+    full_task_status s;
+    s.task_status = task->get_status();
+    s.parent_id = task->get_parent_id();
+    s.abortable = task->is_abortable();
+    s.module = task->get_module();
+    return task->get_progress().then([s] (task_manager::task::progress progress) mutable {
+        s.task_status.progress.completed = progress.completed;
+        s.task_status.progress.total = progress.total;
+        return make_ready_future<json::json_return_type>(make_status(s));
+    });
+}
+
+void set_task_manager(http_context& ctx, routes& r) {
+    tm::get_modules.set(r, [&ctx] (std::unique_ptr<request> req) {
+        return make_ready_future<json::json_return_type>(ctx.tm.local().get_modules());
+    });
+
+    tm::get_tasks.set(r, [&ctx] (std::unique_ptr<request> req) {
+        return do_with(req.get(), [&ctx] (auto &req) {
+            return ctx.tm.map_reduce0([&req] (task_manager& tm) {
+                std::vector<tm::task_stats> res{};
+                std::string m = req->param["module"];
+                if (tm.get_tasks_per_module().contains(m)) {
```

Comment:
We should first lookup the module in the task_manager (it can export a `const std::unordered_map<std:;string, module_ptr>& get_modules()`)
and then use the module to list its tasks.